<a href="https://colab.research.google.com/github/firman-hoerulloh/Python_Association_Analysis_FP-Growth_Algorithm/blob/main/Kelompok_11_Project_Based_Assignment_CLO_1_(Association_Rules_Mining).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project-Based Assignment CLO-1
### Implementasi Analisa Asosiasi

Anggota Kelompok 11:
*   Firman Hoerulloh (1301213392)
*   Arrizal Aryasatya Rizqullah (1301213175)
*   Raihan Kahfi Ananta (1301213463)
*   Ariiq Afrahtama (1301213449)






# Data Preprocessing dan EDA

In [ ]:
import pandas as pd

# Membaca file Excel
df = pd.read_excel('/content/Assignment-1_Data.xlsx')

# Menampilkan 5 baris pertama dari DataFrame
print(df.head())


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


   BillNo                             Itemname  Quantity                Date  \
0  536365   WHITE HANGING HEART T-LIGHT HOLDER         6 2010-12-01 08:26:00   
1  536365                  WHITE METAL LANTERN         6 2010-12-01 08:26:00   
2  536365       CREAM CUPID HEARTS COAT HANGER         8 2010-12-01 08:26:00   
3  536365  KNITTED UNION FLAG HOT WATER BOTTLE         6 2010-12-01 08:26:00   
4  536365       RED WOOLLY HOTTIE WHITE HEART.         6 2010-12-01 08:26:00   

   Price  CustomerID         Country  
0   2.55     17850.0  United Kingdom  
1   3.39     17850.0  United Kingdom  
2   2.75     17850.0  United Kingdom  
3   3.39     17850.0  United Kingdom  
4   3.39     17850.0  United Kingdom  


In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 522064 entries, 0 to 522063
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   BillNo      522064 non-null  object        
 1   Itemname    520609 non-null  object        
 2   Quantity    522064 non-null  int64         
 3   Date        522064 non-null  datetime64[ns]
 4   Price       522064 non-null  float64       
 5   CustomerID  388023 non-null  float64       
 6   Country     522064 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 27.9+ MB


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Diketahui bahwa dataset terdiri dari 7 kolom dengan total 522.064 entri.

In [ ]:
# Periksa nilai yang hilang
print(df.isnull().sum())

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


BillNo             0
Itemname        1455
Quantity           0
Date               0
Price              0
CustomerID    134041
Country            0
dtype: int64


Diketahui terdapat 1455 nilai null pada kolom Itemname dan 134.041 nilai null pada kolom CostumerID

In [ ]:
# Menghapus baris yang memiliki nilai null pada kolom "Itemname"
df1 = df.dropna(subset=['Itemname'])


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Mendapatkan nilai modus dari kolom "CustomerID"
mode_customer_id = df1['CustomerID'].mode()[0]

# Mengisi nilai null pada kolom "CustomerID" dengan nilai modus
df1['CustomerID'] = df1['CustomerID'].fillna(mode_customer_id)

# Menampilkan informasi setelah mengisi nilai null
print(df1.info())



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-6-2ab747cdfdbf>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['CustomerID'] = df1['CustomerID'].fillna(mode_customer_id)


<class 'pandas.core.frame.DataFrame'>
Index: 520609 entries, 0 to 522063
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   BillNo      520609 non-null  object        
 1   Itemname    520609 non-null  object        
 2   Quantity    520609 non-null  int64         
 3   Date        520609 non-null  datetime64[ns]
 4   Price       520609 non-null  float64       
 5   CustomerID  520609 non-null  float64       
 6   Country     520609 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 31.8+ MB
None


In [ ]:
# Periksa nilai yang hilang
print(df1.isnull().sum())

BillNo        0
Itemname      0
Quantity      0
Date          0
Price         0
CustomerID    0
Country       0
dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Memeriksa tipe data dari setiap kolom
df1.dtypes

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


BillNo                object
Itemname              object
Quantity               int64
Date          datetime64[ns]
Price                float64
CustomerID           float64
Country               object
dtype: object

# Model Baseline

In [ ]:
# Mengelompokkan data berdasarkan nomor transaksi (BillNo) dan
# menggabungkan item-item yang terkait dengan setiap transaksi
transactions = df1.groupby('BillNo')['Itemname'].apply(list)

# Menampilkan lima transaksi pertama
print(transactions.head())

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


BillNo
536365    [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
536366    [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
536367    [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
536368    [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
536369                           [BATH BUILDING BLOCK WORD]
Name: Itemname, dtype: object


In [ ]:
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import association_rules

# Membuat one-hot encoding
one_hot = transactions.str.join('|').str.get_dummies()

# Menerapkan algoritma FP-Growth untuk mendapatkan itemset frekuensi tinggi
frequent_itemsets = fpgrowth(one_hot, min_support=0.02, use_colnames=True)

# Membangun aturan asosiasi
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

# Menampilkan aturan asosiasi yang dihasilkan
print("Aturan Asosiasi:")
print(rules)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


Aturan Asosiasi:
                             antecedents                          consequents  \
0             (JAM MAKING SET WITH JARS)             (JAM MAKING SET PRINTED)   
1               (JAM MAKING SET PRINTED)           (JAM MAKING SET WITH JARS)   
2           (ALARM CLOCK BAKELIKE GREEN)           (ALARM CLOCK BAKELIKE RED)   
3             (ALARM CLOCK BAKELIKE RED)         (ALARM CLOCK BAKELIKE GREEN)   
4            (ALARM CLOCK BAKELIKE PINK)           (ALARM CLOCK BAKELIKE RED)   
..                                   ...                                  ...   
159                     (SPOTTY BUNTING)                      (PARTY BUNTING)   
160            (JUMBO BAG RED RETROSPOT)             (JUMBO BAG VINTAGE LEAF)   
161             (JUMBO BAG VINTAGE LEAF)            (JUMBO BAG RED RETROSPOT)   
162  (GARDENERS KNEELING PAD CUP OF TEA)   (GARDENERS KNEELING PAD KEEP CALM)   
163   (GARDENERS KNEELING PAD KEEP CALM)  (GARDENERS KNEELING PAD CUP OF TEA)   

     antec

keterangan Aturan Asosiasi:

*   antecedents: Item atau kelompok item pada sisi kiri aturan.
*   consequents: Item atau kelompok item pada sisi kanan aturan.
*   antecedent support: Persentase transaksi yang mengandung item atau kelompok item pada sisi kiri aturan.
*   consequent support: Persentase transaksi yang mengandung item atau kelompok item pada sisi kanan aturan.
*   support: Persentase transaksi yang mengandung kedua item atau kelompok item pada aturan tersebut.
*   confidence: Tingkat keyakinan bahwa sisi kanan aturan akan muncul dalam transaksi yang mengandung sisi kiri aturan.
*   lift: Pengukuran seberapa besar keterkaitan antara sisi kiri dan sisi kanan aturan. Nilai lift di atas 1 menunjukkan bahwa kedua item tersebut lebih sering muncul bersama daripada yang diharapkan secara acak.
*   leverage: Pengukuran seberapa banyak aturan tersebut "mendorong" penjualan relatif terhadap apa yang diharapkan secara acak.
*   conviction: Pengukuran seberapa kuat hubungan antara sisi kiri dan sisi kanan aturan. Nilai yang lebih tinggi menunjukkan hubungan yang lebih kuat.
*   zhangs_metric: Metrik tambahan untuk mengukur keterkaitan antara item atau kelompok item.

# Analisis Hasil

In [ ]:
# Analisis Hasil
# Menampilkan beberapa statistik penting tentang aturan asosiasi yang dihasilkan
print("\nStatistik Aturan Asosiasi:")
print(rules.describe())


Statistik Aturan Asosiasi:
       antecedent support  consequent support     support  confidence  \
count          164.000000          164.000000  164.000000  164.000000   
mean             0.058911            0.058911    0.024407    0.458246   
std              0.021194            0.021194    0.003985    0.147752   
min              0.025435            0.025435    0.020042    0.183924   
25%              0.044537            0.044537    0.021031    0.360031   
50%              0.053345            0.053345    0.023827    0.442043   
75%              0.062352            0.062352    0.026178    0.552703   
max              0.108967            0.108967    0.040529    0.903509   

             lift    leverage  conviction  zhangs_metric  
count  164.000000  164.000000  164.000000     164.000000  
mean     8.473827    0.021023    1.945430       0.913085  
std      3.634521    0.003970    0.958085       0.062231  
min      1.965254    0.010743    1.123348       0.547032  
25%      6.184674  

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Berikut adalah kesimpulan yang dapat diambil dari output tersebut:

1. Antecedent Support dan Consequent Support:
Rata-rata dari support (persentase transaksi yang mengandung item atau kelompok item pada sisi kiri atau kanan aturan) adalah sekitar 5.89%.
Standar deviasi support adalah sekitar 2.12%, menunjukkan variasi yang cukup signifikan dari nilai-nilai support.
2. Support:
Rata-rata dari support aturan-aturan adalah sekitar 2.44% dari seluruh transaksi.
Standar deviasi dari support adalah sekitar 0.40%, menunjukkan variasi yang cukup besar dalam dukungan aturan-aturan.
3. Confidence:
Rata-rata confidence (tingkat keyakinan bahwa sisi kanan aturan akan muncul dalam transaksi yang mengandung sisi kiri aturan) adalah sekitar 45.82%.
Standar deviasi dari confidence adalah sekitar 14.78%, menunjukkan variasi yang signifikan dalam tingkat keyakinan.
4. Lift:
Rata-rata dari lift (pengukuran seberapa besar keterkaitan antara sisi kiri dan sisi kanan aturan) adalah sekitar 8.47.
Standar deviasi dari lift adalah sekitar 3.63, menunjukkan variasi yang cukup besar dalam tingkat keterkaitan.
5. Leverage:
Rata-rata leverage (pengukuran seberapa banyak aturan tersebut "mendorong" penjualan relatif terhadap apa yang diharapkan secara acak) adalah sekitar 0.021.
Standar deviasi dari leverage adalah sekitar 0.004, menunjukkan variasi yang cukup kecil dalam leverage.
6. Conviction:
Rata-rata conviction (pengukuran seberapa kuat hubungan antara sisi kiri dan sisi kanan aturan) adalah sekitar 1.95.
Standar deviasi dari conviction adalah sekitar 0.96, menunjukkan variasi yang cukup besar dalam tingkat kekuatan hubungan.
7. Zhang's Metric:
Nilai rata-rata dari Zhang's Metric adalah sekitar 0.91, menunjukkan tingkat keterkaitan yang cukup kuat antara item atau kelompok item dalam aturan-aturan asosiasi.

In [ ]:
# Menampilkan aturan asosiasi dengan lift tertinggi
print("\nAturan Asosiasi dengan Lift Tertinggi:")
print(rules[rules['lift'] == rules['lift'].max()])


Aturan Asosiasi dengan Lift Tertinggi:
                                           antecedents  \
134  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENC...   
135                   (PINK REGENCY TEACUP AND SAUCER)   

                                           consequents  antecedent support  \
134                   (PINK REGENCY TEACUP AND SAUCER)            0.036174   
135  (GREEN REGENCY TEACUP AND SAUCER, ROSES REGENC...            0.036421   

     consequent support   support  confidence       lift  leverage  \
134            0.036421  0.025485    0.704514  19.343514  0.024167   
135            0.036174  0.025485    0.699728  19.343514  0.024167   

     conviction  zhangs_metric  
134    3.261000       0.983894  
135    3.209847       0.984147  


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Aturan tersebut menunjukkan bahwa pelanggan yang membeli "GREEN REGENCY TEACUP AND SAUCER" dan "ROSES REGENCY TEACUP AND SAUCER" memiliki kemungkinan sebesar 70.45% untuk juga membeli "PINK REGENCY TEACUP AND SAUCER". Lift yang tinggi (19.34) menunjukkan bahwa pembelian "PINK REGENCY TEACUP AND SAUCER" lebih mungkin terjadi bersamaan dengan pembelian "GREEN REGENCY TEACUP AND SAUCER" dan "ROSES REGENCY TEACUP AND SAUCER" daripada yang diharapkan secara acak. Leverage yang positif menunjukkan adanya peningkatan dalam pembelian bersamaan, sementara conviction yang tinggi menunjukkan kepercayaan yang kuat dalam kebenaran aturan tersebut. Dengan demikian, aturan tersebut dapat menjadi sasaran strategi pemasaran yang efektif untuk meningkatkan penjualan produk "PINK REGENCY TEACUP AND SAUCER" dengan mengaitkannya dengan produk "GREEN REGENCY TEACUP AND SAUCER" dan "ROSES REGENCY TEACUP AND SAUCER".

# Evaluasi

In [ ]:
# Evaluasi dengan mengubah nilai min_support dari 0.02 menjadi 0.01
frequent_itemsets = fpgrowth(one_hot, min_support=0.01, use_colnames=True)

# Membangun aturan asosiasi
rules2 = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

# Menampilkan aturan asosiasi yang dihasilkan
print(rules2)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


                               antecedents                      consequents  \
0                 (HAND WARMER UNION JACK)         (HAND WARMER OWL DESIGN)   
1                 (HAND WARMER OWL DESIGN)         (HAND WARMER UNION JACK)   
2                 (HAND WARMER UNION JACK)  (HAND WARMER SCOTTY DOG DESIGN)   
3          (HAND WARMER SCOTTY DOG DESIGN)         (HAND WARMER UNION JACK)   
4     (WHITE HANGING HEART T-LIGHT HOLDER)  (ASSORTED COLOUR BIRD ORNAMENT)   
...                                    ...                              ...   
2999          (HOT WATER BOTTLE KEEP CALM)          (LOVE HOT WATER BOTTLE)   
3000          (CHOCOLATE HOT WATER BOTTLE)     (HOT WATER BOTTLE KEEP CALM)   
3001          (HOT WATER BOTTLE KEEP CALM)     (CHOCOLATE HOT WATER BOTTLE)   
3002              (HAND WARMER OWL DESIGN)     (HAND WARMER RED LOVE HEART)   
3003          (HAND WARMER RED LOVE HEART)         (HAND WARMER OWL DESIGN)   

      antecedent support  consequent support   supp

Setelah mengubah nilai min_support dari 0.02 menjadi 0.01, jumlah aturan asosiasi yang dihasilkan meningkat dari 164 menjadi 3004. Hal ini menunjukkan bahwa dengan menurunkan nilai min_support, kita lebih memperbolehkan itemset yang lebih jarang muncul untuk dimasukkan ke dalam analisis. Sebagai hasilnya, jumlah aturan asosiasi yang dihasilkan menjadi lebih banyak.

In [ ]:
# Menampilkan aturan asosiasi dengan lift tertinggi
print("\nAturan Asosiasi dengan Lift Tertinggi:")
print(rules2[rules2['lift'] == rules2['lift'].max()])


Aturan Asosiasi dengan Lift Tertinggi:
                 antecedents             consequents  antecedent support  \
1446  (HERB MARKER ROSEMARY)     (HERB MARKER THYME)             0.01158   
1447     (HERB MARKER THYME)  (HERB MARKER ROSEMARY)             0.01153   

      consequent support   support  confidence       lift  leverage  \
1446             0.01153  0.010738     0.92735  80.428744  0.010605   
1447             0.01158  0.010738     0.93133  80.428744  0.010605   

      conviction  zhangs_metric  
1446   13.605998       0.999136  
1447   14.393872       0.999086  


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


setelah mengubah nilai min_support dari 0.02 menjadi 0.01, kita mendapatkan aturan asosiasi dengan lift tertinggi antara HERB MARKER ROSEMARY dan HERB MARKER THYME. Aturan ini memiliki nilai lift yang sangat tinggi, mencapai 80.43, yang menunjukkan bahwa kemungkinan pembelian HERB MARKER ROSEMARY dan HERB MARKER THYME bersama-sama jauh lebih tinggi dari yang diharapkan secara acak.

Kedua item tersebut memiliki dukungan yang relatif rendah (antecedent support dan consequent support di sekitar 0.01), tetapi memiliki tingkat confidence yang sangat tinggi, mencapai sekitar 0.93. Ini menunjukkan bahwa sebagian besar transaksi yang mencakup HERB MARKER ROSEMARY juga mencakup HERB MARKER THYME, dan sebaliknya.

Kesimpulannya, dengan menurunkan nilai min_support, kita dapat menemukan asosiasi yang lebih spesifik dan mungkin lebih relevan dalam dataset, seperti hubungan antara dua jenis ini.